src/script/standard.cpp

```diff
@@ -89,21 +89,45 @@ static constexpr bool IsSmallInteger(opcodetype opcode)
     return opcode >= OP_1 && opcode <= OP_16;
 }
 
+/** Valid number of keys in a multisig are [1,20]. */
+static constexpr bool IsValidMultisigKeyCount(unsigned int n_keys)
+{
+    return n_keys > 0 && n_keys <= MAX_PUBKEYS_PER_MULTISIG;
+}
+
+static bool GetMultisigKeyCount(const CScript& script, CScript::const_iterator& it, unsigned int& count, opcodetype opcode, valtype data)
+{
+    if (IsSmallInteger(opcode)) {
+        count = CScript::DecodeOP_N(opcode);
+    } else {
+        try {
+            count = CScriptNum(data, true).getint();
+        } catch (const scriptnum_error&) {
+            return false;
+        }
+    }
+    if (!IsValidMultisigKeyCount(count)) return false;
+
+    return true;
+}
+
 static bool MatchMultisig(const CScript& script, unsigned int& required, std::vector<valtype>& pubkeys)
 {
     opcodetype opcode;
     valtype data;
+    unsigned int keys;
+
     CScript::const_iterator it = script.begin();
     if (script.size() < 1 || script.back() != OP_CHECKMULTISIG) return false;
 
-    if (!script.GetOp(it, opcode, data) || !IsSmallInteger(opcode)) return false;
-    required = CScript::DecodeOP_N(opcode);
+    if (!script.GetOp(it, opcode, data) || !GetMultisigKeyCount(script, it, required, opcode, data)) return false;
```

Comment:
Does this change behavior, and if so is it intended? bc of the `else` condition https://github.com/bitcoin/bitcoin/pull/20867/files#diff-a1732fc4dc1526b294a41db3f2d3d5e863ec76e68ee3f73449618551f7a469ebR102 I would think `!GetMultisigKeyCount` evaluates differently than the original `!IsSmallInteger(opcode)`

Reply:
It does, that's actually the feature implemented in this PR: handle up to 20 keys (therefore we need to handle not only small integers (<=16) but also pushed ones).